lib/srv/regular/proxy.go

```diff
@@ -270,12 +270,55 @@ func (t *proxySubsys) proxyToHost(ctx context.Context, ch ssh.Channel, clientSrc
 	if teleportVersion != "" && utils.CheckVersion(teleportVersion, utils.MinIPPropagationVersion) != nil {
 		t.doHandshake(ctx, clientSrcAddr, ch, conn)
 	}
+
+	checkedReader := &checkedPrefixReader{
+		ReadWriteCloser: ch,
+		// SSH connection MUST start with "SSH-2.0" bytes according to https://datatracker.ietf.org/doc/html/rfc4253#section-4.2
+		requiredPrefix: []byte("SSH-2.0"),
+	}
+
 	go func() {
-		t.close(utils.ProxyConn(ctx, ch, conn))
+		t.close(utils.ProxyConn(ctx, checkedReader, conn))
 	}()
 	return nil
 }
 
+// checkedPrefixReader checks that read data has the specified prefix.
+type checkedPrefixReader struct {
+	io.ReadWriteCloser
+
+	requiredPrefix  []byte
+	requiredPointer int
+}
+
+func (c *checkedPrefixReader) Read(b []byte) (int, error) {
+	if len(c.requiredPrefix) == c.requiredPointer {
+		// If pointer reached end of required prefix the check was already successful.
+		return c.ReadWriteCloser.Read(b)
+	} else if c.requiredPointer < 0 {
+		// If it's negative, it means check has already failed
+		return 0, trace.AccessDenied("required prefix %q was not found", c.requiredPrefix)
+	}
+
+	n, err := c.ReadWriteCloser.Read(b)
+
+	// Decide which is smaller, read data or remaining portion of the required prefix
+	small, big := c.requiredPrefix[c.requiredPointer:], b[:n]
+	if len(small) > len(big) {
+		big, small = small, big
+	}
+
+	if !bytes.HasPrefix(big, small) {
+		c.requiredPointer = -1
+		return n, trace.AccessDenied("required prefix %q was not found", c.requiredPrefix)
```

Comment:
We should not allow the invalid data "out" of the function (idk if we need to clear the data from the buffer, I reckon it's fine to just say it's unspecified).

```suggestion
		return 0, trace.AccessDenied("required prefix %q was not found", c.requiredPrefix)
```